addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
-                            logger.debug("fadeState: {}", fadeState);
+        public void run() {
+            lock.lock();
+            if (fadeStepper <= steps) {
+                currentFadeState = fromState.fade(toState, (double) fadeStepper / steps);
+                lock.unlock();
 
-                            sendLEDData(fadeState, outputStream);
+                logger.debug("currentFadeState: {}", currentFadeState);
+                if (!ledSender.apply(currentFadeState)) {
+                    logger.warn("Failed sending at step {}", fadeStepper);
+                    throw new RuntimeException("Failed sending at step " + fadeStepper);
+                }
+                lastComTime = System.currentTimeMillis();
+            } else {
+                lock.unlock();
+                if (lastComTime < (System.currentTimeMillis() - keepCommOpenForMS)) {
+                    throw new RuntimeException("Reached end step");
+                }
+            }
+            fadeStepper++;
+        }
+    }
 
-                            busySleep(fadeDurationInMs / fadeSteps, lastTime);
+    synchronized private void fadeToState(final InternalLedState newTargetState) throws IOException {
+
+        if (ledUpdateSyncSemaphore.tryAcquire(1)) {
+            final Future<?> future;
+            final Socket socket;
+
+            // Create and Execute a new LED Fader
+            try {
+                socket = new Socket(host, port);
+                socket.setSoTimeout(DEFAULT_SOCKET_TIMEOUT);
+                final DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream());
+                logger.debug("Connected to '{}'", socket);
+
+                // ensure controller is on
+                sendRaw(getBytesForPower(true), outputStream);
+
+                ledfaderThread = new LEDFaderRunner(
+                        currentFaderState,
+                        newTargetState,
+                        fadeSteps,
+                        KEEPCOMMOPENINMS,
+                        (fs) -> {
+                            try {
+                                sendLEDData(fs, outputStream);
+                                currentFaderState = fs;
+                                // logger.info("Current: {} {} {} {}", fs.getR(), fs.getG(), fs.getB(), fs.getWhite());
+                                return true;
+                            } catch (IOException e) {
+                                logger.warn("IOException", e);
+                                return false;
+                            }
+                        });
+                future = faderFuture.scheduleAtFixedRate(ledfaderThread, 0, fadeDurationInMs / fadeSteps, TimeUnit.MILLISECONDS);
+
+                // Wait untill LED Thread has finished, when so cleanup socket and remove ledfaderThread
+                waiterFuture.schedule(() -> {
+                    try {
+                        future.get();
+                    } catch (Exception e) {
```

Comment:
Given a separate thread will by default otherwise not log its okay to catch that exceptions, I think the change above make it more clear.